lib/bolt/inventory/group.rb

```diff
@@ -129,7 +128,7 @@ def data_merge(data1, data2)
           # Shallow merge instead of deep merge so that vars with a hash value
           # are assigned a new hash, rather than merging the existing value
           # with the value meant to replace it
-          'vars'   => data2['vars'].merge(data1['vars']),
+          'vars'   => data1['vars'].merge(data2['vars']),
```

Comment:
Why was this reversed?

Reply:
Ah, I see the commit references a bug here.

Reply:
I should update the commit message to mention what the bug was...

Reply:
This may have been intentionally reversed, so that `vars` from the 1st entry was the one used. How does it now behave when a node exists in multiple unrelated groups and declares vars in each?

Reply:
Well, the tests there look reasonable and we didn't have tests before. So this behavior seems sane.

Reply:
And the bug appears to be that we prefered group vars over node vars, which definitely doesn't seem right.

Reply:
This method is used in a couple of places.

When we're reducing over the *list* of groups, we pass each group as the first argument, and the accumulator as the *second* argument, which means the first group that sets something wins (it goes into the accumulator and overrides later groups).

When we're traversing *down* the group hierarchy, we pass the parent group as the first argument and the aggregated value from the children as the second argument, so that children win over parents.

When we're combining node-specific data and group-specific data, we pass the group data as the first argument and the node data as the second argument, so that node-specific data wins.

I think that last one means that in this case:

```yaml
groups:
  - name: parent
    nodes:
      - node1:
          vars:
            foo: bar
  groups:
    - name: child
      nodes: node1
      vars:
        foo: baz
```

The node `node1` has vars `foo: bar`. That is, even though the node-specific value is set at the `parent` group level, it wins over the group value in the *more specific* `child` group.

So the total preference order is:
- depth-first search for node-specific values
- depth-first search for group-specific values